module/VuFind/src/VuFind/Recommend/AbstractSearchObjectDeferred.php

```diff
@@ -0,0 +1,169 @@
+ * a particular search object).
+ *
+ * This class sets up an AJAX call to trigger a call to some SearchObject implementation.
+ *
+ * @category VuFind
+ * @package  Recommendations
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @author   Maccabee Levine <msl321@lehigh.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development:plugins:recommendation_modules Wiki
+ */
+abstract class AbstractSearchObjectDeferred implements RecommendInterface
+{
+    /**
+     * Raw configuration parameters
+     *
+     * @var string
+     */
+    protected $rawParams;
+
+    /**
+     * Current search query
+     *
+     * @var string
+     */
+    protected $lookfor;
+
+    /**
+     * Configuration parameters processed for submission via AJAX
+     *
+     * @var string
+     */
+    protected $processedParams;
+
+    /**
+     * Number of expected module parameters (from .ini config)
+     *
+     * @var int
+     */
+    protected $paramCount = 2;
+
+    /**
+     * Store the configuration of the recommendation module.
+     *
+     * @return string Module name in call to AjaxHandler
+     */
+    abstract protected function getAjaxModule();
+
+    /**
+     * Store the configuration of the recommendation module.
+     *
+     * @param string $settings Settings from searches.ini.
+     *
+     * @return void
+     */
+    public function setConfig($settings)
+    {
+        $this->rawParams = $settings;
+    }
+
+    /**
+     * Called before the Search Results object performs its main search
+     * (specifically, in response to \VuFind\Search\SearchRunner::EVENT_CONFIGURED).
+     * This method is responsible for setting search parameters needed by the
+     * recommendation module and for reading any existing search parameters that may
+     * be needed.
+     *
+     * @param \VuFind\Search\Base\Params $params  Search parameter object
+     * @param \Laminas\Stdlib\Parameters $request Parameter object representing user
+     * request.
+     *
+     * @return void
+     */
+    public function init($params, $request)
+    {
+        // Parse out parameters:
+        $settings = explode(':', $this->rawParams);
+
+        // Make sure all elements of the params array are filled in, even if just
+        // with a blank string, so we can rebuild the parameters to pass through
+        // AJAX later on!
+        for ($i = 0; $i < $this->paramCount; $i++) {
+            $settings[$i] ??= '';
+        }
+
+        $this->initLookFor($params, $request, $settings);
+
+        $this->processedParams = implode(':', $settings);
+    }
+
+    /**
+     * Initialize the lookFor query parameter.  Called from init().
+     *
+     * @param \VuFind\Search\Base\Params $params   Search parameter object
+     * @param \Laminas\Stdlib\Parameters $request  Parameter object representing user
+     * request.
+     * @param array                      $settings Parameter array
```

Comment:
I think you may want to add an `&` to the doc comment here to clarify that this is passed by reference (though maybe I'm mistaken and the style checker won't like that). If you can't put the `&` in, I'd at least add "(passed by reference)" to the description to make this clear, because it's an important point!

It might also be worth putting in `use` statements at the top of the file for these classes so you don't have to use the fully qualified names in the doc comments. That might improve readability (though no pressure to change it if it feels like a hassle).

Reply:
An alternative to consider could be having initLookFor return $settings, and then you avoid some of the possible pitfalls/confusion of pass-by-reference by making the data flow a little more explicit. Again, no pressure to change, just an option to consider.

Reply:
Good call about '&' but that's [not supported](https://docs.phpdoc.org/guide/references/phpdoc/tags/param.html#param) (and php-cs-console complains) so I just added the "passed by reference" text instead.

Yeah I considered just returning $settings, but I thought semantically that was wrong, as it's not the *point* of initLookFor(), and some implementations (such as EuropeanaResultsDeferred) don't actually touch it.  The side effect semantics seemed more correct.

I didn't add those 'use' statements, as this followed the (not great) pattern used for init() itself, as well as process().  So as usual I preferred to be consistent if maybe not what's best.

Reply:
Perfect, thanks! No problem about leaving the use statements out for now. I imagine one of these days one of our code standard tools will just decide to do that job automatically for us. ;-)